File: src/csharp/Foundgine.Core/Semantic/Planning/SecurityObligationProof.cs
using Foundgine.Core.Semantic.Security;

namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Machine-readable evidence that every security obligation declared by a
/// rewrite rule was evaluated against the source and rewritten plans.
/// An obligation is either present in the source contract and must remain
/// present, or it is not required by the source contract. The latter is kept
/// explicit so the proof never pretends that an unenforced obligation was
/// actually exercised.
/// </summary>
public sealed record SecurityObligationProof(
    IReadOnlyList<string> Obligations,
    IReadOnlyList<string> Preserved,
    IReadOnlyList<string> NotRequired,
    IReadOnlyList<string> Violations)
{
    public bool IsSatisfied => Violations.Count == 0 &&
                               Obligations.OrderBy(x => x, StringComparer.Ordinal)
                                   .SequenceEqual(
                                       Preserved.Concat(NotRequired).OrderBy(x => x, StringComparer.Ordinal),
                                       StringComparer.Ordinal);

    public static SecurityObligationProof Create(
        IPlanRewriteRule rule,
        SemanticPlan before,
        SemanticPlan after)
    {
        ArgumentNullException.ThrowIfNull(rule);
        ArgumentNullException.ThrowIfNull(before);
        ArgumentNullException.ThrowIfNull(after);

        var obligations = rule.SecurityObligations
            .Distinct(StringComparer.Ordinal)
            .OrderBy(x => x, StringComparer.Ordinal)
            .ToArray();

        var violations = new List<string>();
        foreach (var obligation in obligations)
        {
            if (!SecurityInvariantRegistry.Contains(obligation))
            {
                violations.Add(
                    $"Rewrite rule '{rule.Name}' declares unknown security obligation '{obligation}'.");
            }
        }

        var beforeSet = before.EffectiveSecurityInvariants.ToHashSet(StringComparer.Ordinal);
        var afterSet = after.EffectiveSecurityInvariants.ToHashSet(StringComparer.Ordinal);
        var preserved = new List<string>();
        var notRequired = new List<string>();

        foreach (var obligation in obligations)
        {
            if (!SecurityInvariantRegistry.Contains(obligation))
                continue;

            if (!beforeSet.Contains(obligation))
            {
                notRequired.Add(obligation);
                continue;
            }

            if (!afterSet.Contains(obligation))
            {
                violations.Add(
                    $"Rewrite rule '{rule.Name}' dropped declared security obligation '{obligation}'.");
                continue;
            }

            preserved.Add(obligation);
        }

        var proof = new SecurityObligationProof(obligations, preserved, notRequired, violations);
        if (!proof.IsSatisfied)
            throw new InvalidOperationException(string.Join(" ", violations));

        return proof;
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/SecurityObligationProof.java
package com.foundgine.core.semantic.planning;

import com.foundgine.core.semantic.security.SecurityInvariantRegistry;
import java.util.*;

/** Machine-readable evidence that every security obligation declared by a rewrite rule was evaluated. */
public record SecurityObligationProof(List<String> obligations,List<String> preserved,List<String> notRequired,List<String> violations) {
    public SecurityObligationProof {obligations=List.copyOf(obligations);preserved=List.copyOf(preserved);notRequired=List.copyOf(notRequired);violations=List.copyOf(violations);}
    public boolean isSatisfied(){var left=obligations.stream().sorted().toList();var right=new ArrayList<String>();right.addAll(preserved);right.addAll(notRequired);right.sort(Comparator.naturalOrder());return violations.isEmpty()&&left.equals(right);}
    public static SecurityObligationProof create(IPlanRewriteRule rule,SemanticPlan before,SemanticPlan after){
        Objects.requireNonNull(rule);Objects.requireNonNull(before);Objects.requireNonNull(after);var obligations=rule.securityObligations().stream().distinct().sorted().toList();var violations=new ArrayList<String>();for(var obligation:obligations)if(!SecurityInvariantRegistry.contains(obligation))violations.add("Rewrite rule '"+rule.name()+"' declares unknown security obligation '"+obligation+"'.");var beforeSet=new HashSet<>(before.effectiveSecurityInvariants());var afterSet=new HashSet<>(after.effectiveSecurityInvariants());var preserved=new ArrayList<String>();var notRequired=new ArrayList<String>();for(var obligation:obligations){if(!SecurityInvariantRegistry.contains(obligation))continue;if(!beforeSet.contains(obligation)){notRequired.add(obligation);continue;}if(!afterSet.contains(obligation)){violations.add("Rewrite rule '"+rule.name()+"' dropped declared security obligation '"+obligation+"'.");continue;}preserved.add(obligation);}var proof=new SecurityObligationProof(obligations,preserved,notRequired,violations);if(!proof.isSatisfied())throw new IllegalStateException(String.join(" ",violations));return proof;
    }
}
